### PR TITLE
Add `ceph.rbd.features`

### PR DESCRIPTION
**Pull request by stgraber**

### CODE REVIEW

File: lxd/storage/drivers/driver_ceph_utils.go [hunk 1]
```diff
@@ -63,8 +63,8 @@ func (d *ceph) osdDeletePool() error {
 }
 
 // rbdCreateVolume creates an RBD storage volume.
-// Note that the set of features is intentionally limited is intentionally
-// limited by passing --image-feature explicitly. This is done to ensure that
+// Note that the efault set of features is intentionally limited
```
Comment (tomponline) [inline lxd/storage/drivers/driver_ceph_utils.go:66]: Typo here

**Comment by tomponline** [comment]
Merge conflicts too

File: lxd/storage/drivers/driver_ceph_utils.go (hunk 1 above)
Comment (stgraber) [reply lxd/storage/drivers/driver_ceph_utils.go:66]: fixed

**Comment by tomponline** [comment]
@stgraber for my own info, what does this PR provide, what do ceph features provide?

Also, should we validate them or should we leave that to the ceph tool to do at run time?

**Comment by tomponline** [comment]
And also do we need to persist the features used to create a volume into the volume config so they can be used on future operations on that volume (if the main storage pool feature list changes)?

**Comment by stgraber** [comment]
@tomponline there are a set of rbd features that one can use, things like layering, exclusive-lock, deep-flatten, fast-diff, ...
Exactly what features can be used vary based on the Ceph server version, the Ceph client version and the kernel version.

When we first implemented support, the only safe feature was `layering` so we forced LXD to only ever use that.
This allows for snapshots and cloning but that's about it. Exclusive map is useful to avoid two servers mapping the same volume and deep-flatten/fast-diff allows for more space efficient storage and significantly faster operations like `rbd du`.

Some features can be enabled down the line but some can only be set once at creation of the volume.
For LXD, it doesn't really matter what options were used at creation time so we don't need to record it in the volume.
In fact, when transferring to a remote server, you specifically want to respect its configuration rather than attempt to conserve the source.

That's because attempting to use features that are supported by ceph but not by your kernel will at best result in a failure to `rbd map` and at worst (as the comment implied), a kernel panic.

**Comment by stgraber** [comment]
I noticed this issue because my own cluster is running a very recent 5.8 kernel which supports most of those features in krbd now.
I've enabled them in ceph.conf and manually enabled them on all existing volumes (for those that can be enabled after initial creation) but then noticed that newer volumes were back to just having `layering` enabled.

With this config in place, I can now configure my pools with the set of features I want and update that config should new features become available with more recent kernels.

**Comment by tomponline** [comment]
thanks for the info @stgraber all sounds good.

**Comment by janpekar** [comment]
I think, that enabling fast-diff as default feature is good idea. I noticed, that rebooting lxd server starts a lot of  ```rbd du``` operations which are slow when there is a lot of snapshots.

It can also help to create diffs between rbd images and snapshots (for diff copying) etc.

These features are safe and overhead is not so big in contrast to the benefits for users.

My suggestion is to use 
```
  ceph.rbd.features: layering,exclusive-lock,object-map,fast-diff,deep-flatten
```

Should I create issue for that or start discussion somewhere?
Thank you

**Comment by janpekar** [comment]
Now I got the point of not enabling that features.. if you are using kernel rbd mapping, you need to have at least kernel 5.3. to be able to mount images

https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/commit/?id=d9b9c893048e9d308a833619f0866f1f52778cf5

That is true for 20.04 but not for 18.04...

**Pull request merged**